Add a test skipping option

daniel-chambers opened the pull request:
While Sandwich has the ability to filter a list of tests, sometimes it's very useful to do the inverse and skip tests instead. 
For example, if you know a particular test is temporarily broken, you can skip it in CI, fix it later and then remove the skip once it is fixed.

This PR adds a `--skip` command line option that can be used to pass names of test to skip, in a similar fashion to `--filter`. This command line option is plumbed through a new field on `Options`, which is then consumed in `startSandwichTree'` in order to run a new `SkipTree` interpreter that skips nodes (and their children) if their name contains any of the specified skip strings.

I hope I've updated all the necessary things that require it! 🙂

daniel-chambers commented:
@thomasjm It would be great to get your thoughts on this PR, when you have a chance. 🙂

thomasjm commented:
Hey, sorry for the delay. In general I think it looks great. 

My main thought was that I want the options to be clear about the semantics of `--skip` and `--filter`, especially if they're used together. In retrospect I kind of wish I'd called it `--include` instead of `--filter` since the latter is bit more ambiguous about whether it includes or excludes things. Maybe it would be good to add two aliases: `--include` as an alias for `--filter`, and `--exclude` as an alias for `--skip`?

Currently this PR applies the filter interpreter first, and then the skip one. There's not any situation where applying these in the opposite order would produce a different result is there?

daniel-chambers commented:
Thanks for the reply @thomasjm!

I've run a few scenarios in my head and I haven't come up with an example where I can see reversing the order of filter and skip would change the outcome.

Having said that, I can see that if we change the terminology to "include" and "exclude", the ordering of command-line arguments might imply a behaviour that we don't have. For example, given the following tree:

```
- Test Parent
  - Test Child A
    - Test Grandchild A
    - Test Grandchild B
  - Test Child B
    - Test Grandchild C
```

would the following command line `--exclude "Test Child A" --include "Test Grandchild A"` imply this resultant tree (ie that include can _add back_ things previously excluded)?

```
- Test Parent
  - Test Child A
    - Test Grandchild A
  - Test Child B
    - Test Grandchild C
```

I feel like `--skip "Test Child A" --filter "Test Grandchild A"` doesn't imply "adding back" things, since both skipping and filtering read like subtractive terms to me, whereas "include" reads like an additive term. The behaviour of both FilterTree and SkipTree is subtractive only (the tree only ever gets smaller), so the terms seem to more accurately describe our current behaviour.

What do you think?

thomasjm commented:
Oh I'm not suggesting changing any behavior. I like the current behavior of subtractive functions that get applied sequentially, and are commutative.

FWIW, I wondered what other tools do and decided to look up pytest: https://docs.pytest.org/en/6.2.x/usage.html#specifying-tests-selecting-tests. They have a little expression language you can pass. If we took pytest's lead, we'd just enrich the set of arguments you can pass to `--filter`, so you could do `--filter "foo and not bar"`, `--filter "foo or bar"`, etc. That would be strictly more powerful than having separate `--filter` and `--skip` flags...

daniel-chambers commented:
That expression language is a bag of worms IMHO. One would need to make sure the syntax wasn't ambiguous, which that pytest syntax would be for Sandwich test names (ie. what if my test name was "foo and not bar"?). So at the very least you'd need string quoting and escaping of quotes (ie. `"my test" and not "this \"other\" test"`), and probably parens for overriding operator precedence (ie. `"my test" and ("other test" or "something else")`). Seems overkill for most uses, I think.

Someone who needs that sort of power in their test filtering probably has some very bespoke requirements, and they may be better served by making it easier for them to write their own filtering strategy and plug it in.

thomasjm commented:
Well, fortunately Haskell is good for building parsers. I think such an expression language could be done with a relatively small amount of attoparsec.

In terms of long-term direction I always intended to extend `--filter` to handle regexes at least. And I like the elegance of having a single CLI argument responsible for all filtering. I don't think "bespoke" is the right word for something that's the main filtering mechanism for pytest, which has massive usage.

In the short term, would it improve your life to make the `--skip` option available? We could potentially merge it before the longer-term filtering plans come to fruition.

daniel-chambers commented:
Yes please, I'd like to merge `--skip` before advanced filtering. 🙂

I think the expression language for `--filter` is orthogonal to the `--skip` feature. The semantics of skip aren't exactly reproducible with filter, even with an expression language. Filter says "discard everything except 'these' tests", whereas skip says "keep everything except 'these' tests". Skip could also benefit from a fancy expression language in the future too.

daniel-chambers commented:
Please do correct me if I'm wrong, because I'm running code in my head, but I think with the way FilterTree works right now, it wouldn't be the same, because FilterTree currently continues searching for matches in subtrees even if the parent node has been not been matched and will include the unmatched parent if any of the children do match.

So for example if we ran `--filter 'not "ParentA"' on this tree:

```
- Root
  - ParentA
    - ChildA
  - ParentB
    - ChildB
```

ParentA would still be included because ChildA is "not ParentA". So you'd end up with this:

```
- Root
  - ParentA
    - ChildA
  - ParentB
    - ChildB
```

Whereas `--skip ParentA` in the current PR will drop ParentA and its subtree, and you'd end up with this:

```
- Root
  - ParentB
    - ChildB
```

thomasjm commented:
Yes, you're right. I've been thinking about what this means. I guess the analogy to pytest fails here because pytest suites are flat, and don't have a notion of an "interior node" where a filter function can decide to skip it entirely, or keep it if a child matches.

The fact that this tripped me up makes me even more concerned that the semantics difference between `--filter` and `--skip` will be highly confusing to users.

Just to understand your use-case better -- is skipping entire interior nodes essential to what you want to do? For example, if your goal in your example above were to avoid running `ChildA`, then doing our hypothetical `--filter "not ChildA"` would work.

If filtering interior nodes is actually required, I'm still sort of hoping that the "ideal UI" can be found, which gives all the power one could want and is still easy to understand. Perhaps continuing along the lines of enriching the `--filter` commands, you could do something like `--filter "not ParentA.*"`.

Again, I'm sympathetic to merging a quick fix if this is actively causing you pain and it would improve your life. But I'm a little wary of merging something that isn't the right design and will have to be deprecated later. What do you think?

daniel-chambers commented:
Apologies for the slow reply; I was taking some time off over the Christmas period. 🙂

> is skipping entire interior nodes essential to what you want to do?

Yes, definitely. Before we ported our test suite to Sandwich, we were using HSpec's skipping to target either specific tests (ie ChildA) or specific groupings of tests (ie. ParentA) and skip running them in CI. This was because we were incrementally fixing an area of code in order to make it pass that test suite, so as we fixed the code, we removed the skips from CI to re-enable now-passing tests. Being able to skip groupings of tests meant not having tens of skip args in our CI scripts and just being able to skip entire broken grouping of tests (which usually correlated to a particular feature) with one `--skip`.

We actually liked using HSpec's pathing feature where you could precisely target tests by pathing them from the root (ie `/Root/ParentA`) which was a less fuzzy match. Sandwich doesn't support this, but that's outside the scope of this PR, and we figure our test node names are unique enough that we're hoping it won't matter as a practical matter. I'm only mentioning it here so you've got a full sense of our particular use case.

Here's an example of what our test CI command looked like:

```
  tests-dc-api -- test --agent-base-url "http://localhost:$SQLITE_PORT" \
    --agent-config '{ "db": "/db.chinook.sqlite", "explicit_main_schema": true }' --table-name-prefix '["main"]' \
    --skip "/query API/Order By in Queries/involving relationships"
```

There are many tests under "involving relationships", and the one `--skip` skipped them all. Once we supported relationships in query order bys we removed that `--skip` and let those tests run in CI.

thomasjm commented:
Got it, thanks for explaining! Just wanted to let you know I'm thinking about this, I'm a bit busy right now but I'll try to finish it as soon as possible.

daniel-chambers commented:
@thomasjm Any plans to merge this PR or any further feedback? We're just running into a scenario now where we need to skip a specific test in CI, and while I can just build a fork with this PR's changes and use it, I'd rather not! I'd much rather ensure this PR is good enough to be merged in so that everyone can benefit. 🙂

thomasjm commented:
> Any plans to merge this PR or any further feedback?

Yes, definitely! Sorry, I've just been very busy lately. How about I get back to you with another round within a week?

daniel-chambers commented:
@thomasjm Thank you, I'm grateful for any time you have to spare. 🙂

thomasjm commented:
Okay, here's the long-delayed update! I thought about it and while I'd like to eventually come up with a fancy UI with pathing and a query language of some kind, I think we can go ahead and merge this as long as we document it well.

Here's what I'm thinking: 
1) Rename the option to `--prune`, because I think that more clearly signals that this option removes entire branches of the tree.
2) Run the pruning first and the filtering second. To my mind it's easier to visualize/explain that way: first we prune off entire branches, then we filter to any tests we want to keep. (I'm still not aware of any scenario where changing the order makes a difference, and possibly one doesn't exist, but this way still seems clearer.)
3) Clearly explain the meaning and order of operations in the helper strings for both `--filter` and `--prune`.
4) Also add a bullet point to the docs in `sandwich-site/docs/command_line.md`

Another reasonable choice for the name could be `--skip-subtree`.

What do you think?

daniel-chambers commented:
That sounds very reasonable to me. I think I prefer `--prune` over `--skip-subtree`; it's more concise.

I'll find some time to make these changes.

daniel-chambers commented:
@thomasjm I have made changes to address your feedback. 🙂

thomasjm commented on sandwich/src/Test/Sandwich/Interpreters/PruneTree.hs:
```diff
@@ -0,0 +1,49 @@
+import qualified Data.List as L
+import Test.Sandwich.Types.Spec
+
+pruneTree :: Free (SpecCommand context m) () -> [String] -> Free (SpecCommand context m) ()
```
It's worth noting that `filterTree` takes `String` instead of `[String]`, so the caller is expected to `foldl` this to handle multiple filters. Might be good to do this the same way for consistency. I kind of like that the fold makes it clear at the callsite that pruning with multiple arguments is applied sequentially. The current `pruneTree` type signature is consistent with wrong interpretations, like that it only prunes nodes that contain *all* of the arguments. Not a huge deal, but I like making the picture as clear as possible.

daniel-chambers replied:
The reason I did it that way was that it seemed more efficient, since the list of prune labels would likely be smaller than the tree, so walking the tree once rather than for each label seemed more efficient. 

However, the performance impact is likely irrelevantly small here, so I've refactored it to your more semantically clear version.

thomasjm commented on sandwich/src/Test/Sandwich/Interpreters/PruneTree.hs:
```diff
@@ -0,0 +1,49 @@
+            (Pure _) -> go next
+            subspec' -> Free (Introduce'' loc no label' cl alloc cleanup subspec' (go next))
+        | otherwise ->
+          go next
```
Nit: the normal style in this project isn't to make a newline for short lines like here and below (and `go tree` above)

daniel-chambers replied:
Done.

thomasjm reviewed (changes requested):
Thanks, here's a round of review!

daniel-chambers commented on sandwich-site/docs/command_line.md:
```diff
@@ -13,7 +13,8 @@ If you use [runSandwichWithCommandLineArgs](http://hackage.haskell.org/package/s
 * Repeat the test suite N times: `--repeat n` (useful to exercise flaky tests)
-* Filter the test tree to a string: `--filter some_string`.
+* Filter the test tree to a string: `--filter some_string`. Filtering happens after pruning, no matter the order of the arguments.
+* Prunes any nodes (and their children) in the test tree whose label contains a particular string: `--skip some_string`. Pruning happens before filtering, no matter the order of the arguments.
```
Oops, fixed.

daniel-chambers commented on sandwich/src/Test/Sandwich/ArgParsing.hs:
```diff
@@ -86,7 +86,8 @@ mainCommandLineOptions userOptionsParser individualTestParser = CommandLineOptio
   <*> optional (option auto (long "visibility-threshold" <> short 'v' <> showDefault <> help "Set the visibility threshold for formatters" <> metavar "INT"))
-  <*> many (strOption (long "filter" <> short 'f' <> help "Filter test tree by string matching text example labels" <> metavar "STRING"))
+  <*> many (strOption (long "filter" <> short 'f' <> help "Filter test tree by string matching text example labels. Filtering happens after pruning, if any" <> metavar "STRING"))
+  <*> many (strOption (long "prune" <> short 'p' <> help "Prune test subtrees by string matching text example labels. The matched test and all its children are removed. Pruning happens before filtering, if any" <> metavar "STRING"))
```
I've moved everything around.

daniel-chambers commented on sandwich/src/Test/Sandwich/Internal/Running.hs:
```diff
@@ -36,10 +37,11 @@ startSandwichTree options spec = do
   startSandwichTree' baseContext options spec
 
 startSandwichTree' :: BaseContext -> Options -> CoreSpec -> IO [RunNode BaseContext]
```
Done.

daniel-chambers commented:
@thomasjm I've made the requested changes, thank you.

thomasjm commented:
Look great, thanks! I'll cut a release with this soon.

thomasjm commented:
Released in https://hackage.haskell.org/package/sandwich-0.1.3.0, thanks for the contribution!

daniel-chambers commented:
@thomasjm Thank you!
